*** mrt-2.2.2a.orig/src/programs/route_btoa/route_btoa.c	2000-04-28 21:31:14.000000000 +0200
--- mrt-2.2.2a.modified/src/programs/route_btoa/route_btoa.c	2009-10-29 13:50:28.776392476 +0100
***************
*** 12,17 ****
--- 12,19 ----
  #define OUTPUT_MODE_MACHINE2 2	/* masaki uses this */
  #define OUTPUT_MODE_RAW 3
  static int output_mode = OUTPUT_MODE_TEXT;
+ static int as_filter = 0;
+ static int as_filter_enabled = 0;
  static char time_str[MAXLINE];
  
  static io_t *IO;
***************
*** 20,25 ****
--- 22,30 ----
  
  void process_input ();
  
+ static void
+ print_msg_hdr (mrt_msg_t *msg);
+ 
  static void 
  print_bgp_msg (mrt_msg_t * msg)
  {
***************
*** 105,110 ****
--- 110,122 ----
  			      &family, &gateway_from, &gateway_to);
      if (cp == NULL)
  	return;
+     
+     if(as_filter_enabled) {
+ 	if(gateway_from->AS!= as_filter)
+ 	return;
+     
+     	print_msg_hdr (msg);
+     }
  
      if (msg->type == MSG_PROTOCOL_BGP4MP) {
  	assert (msg->subtype == BGP4MP_MESSAGE || 
***************
*** 700,706 ****
      int c;
      extern char *optarg;	/* getopt stuff */ 
      extern int optind;		/* getopt stuff */ 
!     char *usage = "Usage: route_btoa [-m] [(-i|-r)] [input_binary_file(s)] \n"; 
      int errs; 
      int input_mode = 0;
      
--- 712,718 ----
      int c;
      extern char *optarg;	/* getopt stuff */ 
      extern int optind;		/* getopt stuff */ 
!     char *usage = "Usage: route_btoa [-m] [-a AS] [(-i|-r)] [input_binary_file(s)] \n"; 
      int errs; 
      int input_mode = 0;
      
***************
*** 713,721 ****
      io_set (IO, IO_INFILE, "stdin", NULL);
   
  
!     while ((c = getopt (argc, argv, "if:o:mMuvh")) != -1)
        switch (c) {
  	/* machine parseable format */
        case 'u':
  	output_mode = OUTPUT_MODE_RAW;
  	break;
--- 725,737 ----
      io_set (IO, IO_INFILE, "stdin", NULL);
   
  
!     while ((c = getopt (argc, argv, "a:if:o:mMuvh")) != -1)
        switch (c) {
  	/* machine parseable format */
+       case 'a':
+ 	as_filter_enabled = 1;
+ 	as_filter=atoi(optarg);
+ 	break;
        case 'u':
  	output_mode = OUTPUT_MODE_RAW;
  	break;
***************
*** 831,837 ****
--- 847,855 ----
      case MSG_PROTOCOL_BGP4PLUS_01:
  #endif /* HAVE_IPV6 */
      case MSG_PROTOCOL_BGP4MP:
+       if(!as_filter_enabled) {
        print_msg_hdr (msg);
+       } 
        print_bgp_msg (msg);
        break;
      case MSG_TABLE_DUMP:
